github/transport.go

```diff
@@ -29,3 +38,113 @@ func (ett *etagTransport) RoundTrip(req *http.Request) (*http.Response, error) {
 func NewEtagTransport(rt http.RoundTripper) *etagTransport {
 	return &etagTransport{transport: rt}
 }
+
+// rateLimitTransport implements GitHub's best practices
+// for avoiding rate limits
+// https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits
+type rateLimitTransport struct {
+	transport        http.RoundTripper
+	delayNextRequest bool
+	responseBody     []byte
+
+	m sync.Mutex
+}
+
+func (rlt *rateLimitTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Make requests for a single user or client ID serially
+	// This is also necessary for safely saving
+	// and restoring bodies between retries below
+	rlt.lock(req)
+
+	// If you're making a large number of POST, PATCH, PUT, or DELETE requests
+	// for a single user or client ID, wait at least one second between each request.
+	if rlt.delayNextRequest {
+		log.Printf("[DEBUG] Sleeping %s between write operations", writeDelay)
+		time.Sleep(writeDelay)
+	}
+
+	rlt.delayNextRequest = isWriteMethod(req.Method)
+
+	resp, err := rlt.transport.RoundTrip(req)
+	if err != nil {
+		rlt.unlock(req)
+		return resp, err
+	}
+
+	// Make response body accessible for retries & debugging
+	// (work around bug in GitHub SDK)
+	// See https://github.com/google/go-github/pull/986
+	r1, r2, err := drainBody(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	resp.Body = r1
+	ghErr := github.CheckResponse(resp)
```

Comment:
I'm seeing now that `github.CheckResponse` accepts the `response` object, making the use of `TeeReader` impossible. The first time I read it I thought it accepted just a `Reader`, in that case `TeeReader` would have worked nicely
```
var r2 bytes.Buffer
r1 := io.TeeReader(resp.Body, &r2)
ghErr := github.CheckResponse(r1)
resp.Body = &r2
```
Also the interfaces might not have aligned, if `resp.Body` was `ReadCloser` you would have to wrap it just like you did in `drainBody` anyways nevermind me... always looking for an excuse to use `io` builtins 😛